Documentation/ceph-best-practices.md

```diff
@@ -0,0 +1,516 @@
+
+Recommendations for MGR nodes are harder to make, since enabling more modules means higher usage. We
+recommend starting with 2 GB RAM and 2 CPU cores for MGRs. It is a good idea to look at the actual
+usage for Deployments, and do not forget to consider usage during failure scenarios.
+
+MONs:
+ * Request 4 CPU cores
+ * Request 4GB RAM (2.5GB minimum)
+
+MGR:
+ * Memory will grow the more MGR modules are enabled
+ * Request 2 GB RAM and 2 CPU cores
+
+### Resource Requests - OSD CPU
+
+Recommendations and calculations for OSD CPU are straightforward.
+
+Hardware recommendations:
+ * 1 x 2GHz CPU Thread per spinner
+ * 2 x GHz CPU Thread per SSD
+ * 4 x GHz CPU Thread per NVMe
+
+Examples:
+ * 8 HDDS journaled to SSD – `10 cores / 8 OSDs = 1.25 cores per OSD`
+ * 6 SSDs without journals – `12 cores / 6 OSDs = 2 cores per OSD`
+ * 8 SSDs journaled to NVMe – `20 cores / 8 OSDs = 2.5 cores per OSD`
+
+Note that resources are applied cluster-wide to all OSDs. If a cluster contains multiple OSD types,
+you must use the highest Requests for the whole cluster. For the examples below, a mixture of HDDs
+journaled to SSD and SSDs without journals would necessitate a Request for 2 cores.
+
+### Resource Requests - OSD RAM
+
+We have node hardware recommendations for OSD RAM usage, and we need to translate this to RAM
+requests on a per-OSD basis. The node-level recommendation we have is given below where
+`osd_memory_target` is a Ceph configuration described more afterwards.
+
+> `Total RAM required = [number of OSDs] x (1 GB + osd_memory_target) + 16 GB `
+
+Ceph OSDs will attempt to keep heap memory usage under a designated target size set via the
+`osd_memory_target` configuration option. Ceph’s default `osd_memory_target` is 4GB, and it is
+recommended to never decrease the `osd_memory_target` below 4GB. You may wish to increase this value
+to improve overall Ceph read performance by allowing the OSDs to use more RAM. While the total
+amount of heap memory mapped by the process should generally stay close to this target, there is no
+guarantee that the Linux kernel will actually reclaim memory that has been unmapped.
+
+As an example, for a node hosting 8 OSDs, memory Requests would be calculated as below.
+
+> `8 OSDs x (1GB + 4GB) + 16GB = 56GB per node`
+
+This gives us resource usage for each OSD.
+
+> `56GB / 8 OSDs = 7GB`
+
+Ceph has a feature that allows it to set `osd_memory_target` automatically when a Rook OSD Resource
+Request is set; however, Ceph sets this value 1:1 and does not leave overhead for waiting for the
+kernel to free memory. Therefore, we recommend setting `osd_memory_target` in Ceph explicitly and
+setting Rook’s OSD resource requests to a higher value than `osd_memory_target` by at least an
+additional 1GB so that Kubernetes does not schedule more applications or Ceph daemons onto a node
+than the node is likely to have RAM available for.
+
```

Comment:
Holy toledo batman this sentence is 4 lines long. See suggestions above.

Reply:
Again, the original split this out so that it was clear and concise, please consider revising:

_"Therefore, we recommend setting `osd_memory_target` in Ceph explicitly, even if you wish to use the default value. Set Rook’s OSD resource requests accordingly and to a higher value than `osd_memory_target` by at least an additional 1GB. This is so Kubernetes does not schedule more applications or Ceph daemons onto a node than the node is likely to have RAM available for."_

Reply:
I think I took this text from my original draft of it. Definitely revised.